From 4f668d9268075c410f0663d44a5fc79c122377bc Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 Jul 2024 00:37:13 +0200
Subject: [PATCH] Fix signrawtransactionwith{key,wallet} (#30)

Currently, the signatures signed with the RPC aren't valid on Dogecoin. Change the RPC to not use FORKID in legacy mode.

diff --git a/src/rpc/rawtransaction_util.cpp b/src/rpc/rawtransaction_util.cpp
index 4a84a4029f8db5b5f7853876d4f618ed944b9661..f21a44b38e2342fa63e817417e74a3eb49a83965 100644
--- a/src/rpc/rawtransaction_util.cpp
+++ b/src/rpc/rawtransaction_util.cpp
@@ -20,6 +20,20 @@
 #include <util/strencodings.h>
 #include <util/vector.h>
 
+void EnsureCorrectSighash(SigHashType sigHashType) {
+    if (GetSignScriptFlags() & SCRIPT_ENABLE_SIGHASH_FORKID) {
+        if (!sigHashType.hasForkId()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER,
+                               "Signature must use SIGHASH_FORKID");
+        }
+    } else {
+        if (sigHashType.hasForkId()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER,
+                               "Illegal use of SIGHASH_FORKID");
+        }
+    }
+}
+
 CMutableTransaction ConstructTransaction(const CChainParams &params,
                                          const UniValue &inputs_in,
                                          const UniValue &outputs_in,
@@ -265,10 +279,7 @@ void SignTransaction(CMutableTransaction &mtx, const SigningProvider *keystore,
                      const std::map<COutPoint, Coin> &coins,
                      const UniValue &hashType, UniValue &result) {
     SigHashType sigHashType = ParseSighashString(hashType);
-    if (!sigHashType.hasForkId()) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER,
-                           "Signature must use SIGHASH_FORKID");
-    }
+    EnsureCorrectSighash(sigHashType);
 
     // Script verification errors
     std::map<int, std::string> input_errors;
diff --git a/src/rpc/rawtransaction_util.h b/src/rpc/rawtransaction_util.h
index c60ec6a089aca8e8d076e5d131610d874a4e8cfd..e4ab1e4bcf8fc5c0284040f288032e671d685560 100644
--- a/src/rpc/rawtransaction_util.h
+++ b/src/rpc/rawtransaction_util.h
@@ -19,6 +19,10 @@ class Coin;
 class COutPoint;
 class SigningProvider;
 class UniValue;
+class SigHashType;
+
+/** Make sure we're using a compatible SIGHASH */
+void EnsureCorrectSighash(SigHashType sigHashType);
 
 /**
  * Sign a transaction with the given keystore and previous transactions
diff --git a/src/script/sign.cpp b/src/script/sign.cpp
index 0bdb16eca39a795b9b1d92242699e11d1c47baa5..d34039c3710dc295869f571cf065e6d68779fc80 100644
--- a/src/script/sign.cpp
+++ b/src/script/sign.cpp
@@ -5,6 +5,8 @@
 
 #include <script/sign.h>
 
+#include <chainparams.h>
+#include <consensus/activation.h>
 #include <consensus/amount.h>
 #include <key.h>
 #include <policy/policy.h>
@@ -47,6 +49,12 @@ bool MutableTransactionSignatureCreator::CreateSig(
     return true;
 }
 
+uint32_t GetSignScriptFlags() {
+    return IsLegacyScriptRulesEnabled(Params().GetConsensus())
+               ? STANDARD_SCRIPT_VERIFY_FLAGS_LEGACY
+               : STANDARD_SCRIPT_VERIFY_FLAGS;
+}
+
 static bool GetCScript(const SigningProvider &provider,
                        const SignatureData &sigdata, const CScriptID &scriptid,
                        CScript &script) {
@@ -239,7 +247,7 @@ bool ProduceSignature(const SigningProvider &provider,
     // Test solution
     sigdata.complete =
         solved && VerifyScript(sigdata.scriptSig, fromPubKey,
-                               STANDARD_SCRIPT_VERIFY_FLAGS, creator.Checker());
+                               GetSignScriptFlags(), creator.Checker());
     return sigdata.complete;
 }
 
@@ -293,8 +301,8 @@ SignatureData DataFromTransaction(const CMutableTransaction &tx,
     // Get signatures
     MutableTransactionSignatureChecker tx_checker(&tx, nIn, txout.nValue);
     SignatureExtractorChecker extractor_checker(data, tx_checker);
-    if (VerifyScript(data.scriptSig, txout.scriptPubKey,
-                     STANDARD_SCRIPT_VERIFY_FLAGS, extractor_checker)) {
+    if (VerifyScript(data.scriptSig, txout.scriptPubKey, GetSignScriptFlags(),
+                     extractor_checker)) {
         data.complete = true;
         return data;
     }
@@ -328,7 +336,7 @@ SignatureData DataFromTransaction(const CMutableTransaction &tx,
                 // a signature and it is valid
                 if (data.signatures.count(CPubKey(pubkey).GetID()) ||
                     extractor_checker.CheckSig(sig, pubkey, next_script,
-                                               STANDARD_SCRIPT_VERIFY_FLAGS)) {
+                                               GetSignScriptFlags())) {
                     last_success_key = i + 1;
                     break;
                 }
@@ -440,9 +448,8 @@ bool IsSolvable(const SigningProvider &provider, const CScript &script) {
     SignatureData sigs;
     if (ProduceSignature(provider, DUMMY_SIGNATURE_CREATOR, script, sigs)) {
         // VerifyScript check is just defensive, and should never fail.
-        bool verified =
-            VerifyScript(sigs.scriptSig, script, STANDARD_SCRIPT_VERIFY_FLAGS,
-                         DUMMY_CHECKER);
+        bool verified = VerifyScript(sigs.scriptSig, script,
+                                     GetSignScriptFlags(), DUMMY_CHECKER);
         assert(verified);
         return true;
     }
@@ -483,9 +490,9 @@ bool SignTransaction(CMutableTransaction &mtx, const SigningProvider *keystore,
         }
 
         ScriptError serror = ScriptError::OK;
-        if (!VerifyScript(
-                txin.scriptSig, prevPubKey, STANDARD_SCRIPT_VERIFY_FLAGS,
-                TransactionSignatureChecker(&txConst, i, amount), &serror)) {
+        if (!VerifyScript(txin.scriptSig, prevPubKey, GetSignScriptFlags(),
+                          TransactionSignatureChecker(&txConst, i, amount),
+                          &serror)) {
             if (serror == ScriptError::INVALID_STACK_OPERATION) {
                 // Unable to sign input and verification failed (possible
                 // attempt to partially sign).
diff --git a/src/script/sign.h b/src/script/sign.h
index 01f3bc57d0b68e20e43b8dfe3f0d3dc8c3b9ee69..f3f0ae2e5bb743457b9aae10ae3006063b8d3231 100644
--- a/src/script/sign.h
+++ b/src/script/sign.h
@@ -169,6 +169,9 @@ void SerializeHDKeypaths(Stream &s,
     }
 }
 
+/** Get the script_flags to be used for signing */
+uint32_t GetSignScriptFlags();
+
 /** Produce a script signature using a generic signature creator. */
 bool ProduceSignature(const SigningProvider &provider,
                       const BaseSignatureCreator &creator,
diff --git a/src/wallet/rpcwallet.cpp b/src/wallet/rpcwallet.cpp
index 96f13ee656b4c91c113629e7d87837581a10defe..ad560e0ab862a7e76a3d09a1d8aab6d2d5a683c1 100644
--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -3663,10 +3663,7 @@ RPCHelpMan signrawtransactionwithwallet() {
             ParsePrevouts(request.params[1], nullptr, coins);
 
             SigHashType nHashType = ParseSighashString(request.params[2]);
-            if (!nHashType.hasForkId()) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER,
-                                   "Signature must use SIGHASH_FORKID");
-            }
+            EnsureCorrectSighash(nHashType);
 
             // Script verification errors
             std::map<int, std::string> input_errors;
@@ -4603,10 +4600,7 @@ static RPCHelpMan walletprocesspsbt() {
 
             // Get the sighash type
             SigHashType nHashType = ParseSighashString(request.params[2]);
-            if (!nHashType.hasForkId()) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER,
-                                   "Signature must use SIGHASH_FORKID");
-            }
+            EnsureCorrectSighash(nHashType);
 
             // Fill transaction with our data and also sign
             bool sign = request.params[1].isNull()
diff --git a/test/functional/feature_legacy_signrawtransaction.py b/test/functional/feature_legacy_signrawtransaction.py
new file mode 100644
index 0000000000000000000000000000000000000000..d7d495c4fb3e77a8a8112b757e4988495c7d4f90
--- /dev/null
+++ b/test/functional/feature_legacy_signrawtransaction.py
@@ -0,0 +1,185 @@
+# Copyright (c) 2024 The Bitcoin developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test we sign valid Dogecoin legacy signatures for transactions."""
+
+from test_framework.address import ADDRESS_ECREG_UNSPENDABLE
+from test_framework.key import ECKey
+from test_framework.messages import COutPoint, CTransaction, CTxIn, CTxOut
+from test_framework.script import (
+    OP_CHECKSIG,
+    OP_DUP,
+    OP_EQUALVERIFY,
+    OP_HASH160,
+    CScript,
+    hash160,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_raises_rpc_error
+from test_framework.wallet_util import bytes_to_wif
+
+
+class LegacyScriptSignrawtransactionTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 2
+        self.extra_args = [["-legacyscriptrules"], []]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def run_test(self):
+        node = self.nodes[0]
+        nonlegacy_node = self.nodes[1]
+
+        coinblockhash1 = self.generate(node, 1)[0]
+        cointxid1 = node.getblock(coinblockhash1)["tx"][0]
+
+        coinblockhash2 = self.generate(nonlegacy_node, 1)[0]
+        cointxid2 = nonlegacy_node.getblock(coinblockhash2)["tx"][0]
+
+        self.generatetoaddress(node, 99, ADDRESS_ECREG_UNSPENDABLE)
+
+        key = ECKey()
+        key.set(b"ABCD" * 8, True)
+        pubkey = key.get_pubkey().get_bytes()
+        p2pkh = CScript(
+            [OP_DUP, OP_HASH160, hash160(pubkey), OP_EQUALVERIFY, OP_CHECKSIG]
+        )
+
+        coinvalue = 5000000000
+
+        tx = CTransaction()
+        tx.vin = [CTxIn(COutPoint(int(cointxid1, 16), 0))]
+        tx.vout = [CTxOut(coinvalue - 1000, p2pkh)]
+
+        nonlegacy_tx = CTransaction()
+        nonlegacy_tx.vin = [CTxIn(COutPoint(int(cointxid2, 16), 0))]
+        nonlegacy_tx.vout = [CTxOut(coinvalue - 1000, p2pkh)]
+
+        legacy_sighashes = [
+            "ALL",
+            "NONE",
+            "SINGLE",
+            "ALL|ANYONECANPAY",
+            "NONE|ANYONECANPAY",
+            "SINGLE|ANYONECANPAY",
+        ]
+        for sighash in legacy_sighashes:
+            result = node.signrawtransactionwithwallet(tx.serialize().hex(), [], sighash)
+            assert result["complete"]
+            assert_raises_rpc_error(
+                -8,
+                "Signature must use SIGHASH_FORKID",
+                nonlegacy_node.signrawtransactionwithwallet,
+                nonlegacy_tx.serialize().hex(),
+                [],
+                sighash,
+            )
+
+        bip143_sighashes = [
+            "ALL|FORKID",
+            "NONE|FORKID",
+            "SINGLE|FORKID",
+            "ALL|FORKID|ANYONECANPAY",
+            "NONE|FORKID|ANYONECANPAY",
+            "SINGLE|FORKID|ANYONECANPAY",
+        ]
+        for sighash in bip143_sighashes:
+            assert_raises_rpc_error(
+                -8,
+                "Illegal use of SIGHASH_FORKID",
+                node.signrawtransactionwithwallet,
+                tx.serialize().hex(),
+                [],
+                sighash,
+            )
+            result = nonlegacy_node.signrawtransactionwithwallet(
+                nonlegacy_tx.serialize().hex(), [], sighash
+            )
+            assert result["complete"]
+
+        # Legacy tx works
+        signed_tx = node.signrawtransactionwithwallet(tx.serialize().hex(), [], "ALL")
+        assert signed_tx["complete"]
+        txid = node.sendrawtransaction(signed_tx["hex"])
+        assert_raises_rpc_error(
+            -26,
+            "mandatory-script-verify-flag-failed (Signature must use SIGHASH_FORKID)",
+            nonlegacy_node.sendrawtransaction,
+            signed_tx["hex"],
+        )
+
+        signed_tx = nonlegacy_node.signrawtransactionwithwallet(
+            nonlegacy_tx.serialize().hex(), [], "ALL|FORKID"
+        )
+        assert signed_tx["complete"]
+        nonlegacy_txid = nonlegacy_node.sendrawtransaction(signed_tx["hex"])
+        # Note: NULLFAIL is not mandatory on legacy, so we just get a FALSE from OP_CHECKSIG
+        assert_raises_rpc_error(
+            -26,
+            "mandatory-script-verify-flag-failed (Script evaluated without error but finished with a false/empty top stack element)",
+            node.sendrawtransaction,
+            signed_tx["hex"],
+        )
+
+        tx2 = CTransaction()
+        tx2.vin = [CTxIn(COutPoint(int(txid, 16), 0))]
+        tx2.vout = [CTxOut(coinvalue - 2000, p2pkh)]
+
+        nonlegacy_tx2 = CTransaction()
+        nonlegacy_tx2.vin = [CTxIn(COutPoint(int(nonlegacy_txid, 16), 0))]
+        nonlegacy_tx2.vout = [CTxOut(coinvalue - 2000, p2pkh)]
+
+        tx2_keys = [bytes_to_wif(key.get_bytes())]
+        tx2_outpoints = [
+            {
+                "txid": txid,
+                "vout": 0,
+                "scriptPubKey": p2pkh.hex(),
+                "amount": tx.vout[0].nValue / 100,
+            }
+        ]
+        nonlegacy_tx2_outpoints = [
+            {
+                "txid": nonlegacy_txid,
+                "vout": 0,
+                "scriptPubKey": p2pkh.hex(),
+                "amount": tx.vout[0].nValue / 100,
+            }
+        ]
+        for sighash in legacy_sighashes:
+            result = node.signrawtransactionwithkey(
+                tx2.serialize().hex(), tx2_keys, tx2_outpoints, sighash
+            )
+            assert result["complete"]
+            assert_raises_rpc_error(
+                -8,
+                "Signature must use SIGHASH_FORKID",
+                nonlegacy_node.signrawtransactionwithkey,
+                nonlegacy_tx2.serialize().hex(),
+                tx2_keys,
+                nonlegacy_tx2_outpoints,
+                sighash,
+            )
+        for sighash in bip143_sighashes:
+            assert_raises_rpc_error(
+                -8,
+                "Illegal use of SIGHASH_FORKID",
+                node.signrawtransactionwithkey,
+                tx2.serialize().hex(),
+                tx2_keys,
+                tx2_outpoints,
+                sighash,
+            )
+            result = nonlegacy_node.signrawtransactionwithkey(
+                nonlegacy_tx2.serialize().hex(),
+                tx2_keys,
+                nonlegacy_tx2_outpoints,
+                sighash,
+            )
+            assert result["complete"]
+
+
+if __name__ == "__main__":
+    LegacyScriptSignrawtransactionTest().main()
